pkg/jobstore/boltdb/context.go

```diff
@@ -39,14 +41,19 @@ func newTxContext(ctx context.Context, tx *bolt.Tx) *txContext {
 		Context:    innerCtx,
 		tx:         tx,
 		cancelFunc: cancelFunc,
+		closeCh:    make(chan struct{}),
 	}
 	// Start a goroutine that listens for the context's Done channel.
 	go func() {
-		<-innerCtx.Done()
-
-		// Attempt to rollback the transaction, which is a no-op if already committed or rolled back.
-		if err := txCtx.doRollback(); err != nil {
-			log.Ctx(txCtx.Context).Error().Err(err).Msg("failed to rollback transaction on context cancellation")
+		defer func() {
+			// Attempt to rollback the transaction, which is a no-op if already committed or rolled back.
+			if err := txCtx.doRollback(); err != nil {
+				log.Ctx(txCtx).Error().Err(err).Msg("failed to rollback transaction on tx cleanup")
+			}
+		}()
+		select {
+		case <-innerCtx.Done():
+		case <-txCtx.closeCh:
 		}
```

Comment:
We can simplify this by extracting the logic of this deferred function to below the select statement and removing the defer statement.